java/src/main/native/src/row_conversion.cu

```diff
@@ -373,140 +975,634 @@ static int calc_fixed_width_kernel_dims(const cudf::size_type num_columns,
+    auto const row_size_with_end_pad =
+        util::round_up_unsafe(row_size_with_this_col, JCUDF_ROW_ALIGNMENT);
+
+    if (row_size_with_end_pad * window_height > shmem_limit_per_block) {
+      // too large, close this window, generate vertical blocks and restart
+      f(current_window_start_col, col == 0 ? col : col - 1, window_height);
+
+      row_size =
+          util::round_up_unsafe((column_starts[col] + column_sizes[col]) & 7, alignment_needed);
+      row_size += col_size; // alignment required for shared memory window boundary to match
+                            // alignment of output row
+      current_window_start_col = col;
+      current_window_width = 0;
+    } else {
+      row_size = row_size_with_this_col;
+      current_window_width++;
+    }
+  }
+
+  // build last set of blocks
+  if (current_window_width > 0) {
+    f(current_window_start_col, static_cast<int>(column_sizes.size()) - 1, window_height);
+  }
+}
+
+#endif // #if !defined(__CUDA_ARCH__) || __CUDA_ARCH__ >= 700
+
+} // namespace detail
+
+std::vector<std::unique_ptr<column>> convert_to_rows(table_view const &tbl,
+                                                     rmm::cuda_stream_view stream,
+                                                     rmm::mr::device_memory_resource *mr) {
+#if !defined(__CUDA_ARCH__) || __CUDA_ARCH__ >= 700
+  auto const num_columns = tbl.num_columns();
+  auto const num_rows = tbl.num_rows();
+
+  int device_id;
+  CUDA_TRY(cudaGetDevice(&device_id));
+  int total_shmem;
+  CUDA_TRY(cudaDeviceGetAttribute(&total_shmem, cudaDevAttrMaxSharedMemoryPerBlock, device_id));
+
+  // TODO: why is this needed. kernel fails to launch if all memory is requested.
+  total_shmem -= 1024;
+  auto const shmem_limit_per_block = total_shmem / NUM_BLOCKS_PER_KERNEL_LOADED;
+
+  // break up the work into blocks, which are a starting and ending row/col #.
+  // this window size is calculated based on the shared memory size available
+  // we want a single block to fill up the entire shared memory space available
+  // for the transpose-like conversion.
+
+  // There are two different processes going on here. The GPU conversion of the data
+  // and the writing of the data into the list of byte columns that are a maximum of
+  // 2 gigs each due to offset maximum size. The GPU conversion portion has to understand
+  // this limitation because the column must own the data inside and as a result it must be
+  // a distinct allocation for that column. Copying the data into these final buffers would
+  // be prohibitively expensive, so care is taken to ensure the GPU writes to the proper buffer.
+  // The windows are broken at the boundaries of specific rows based on the row sizes up
+  // to that point. These are row batches and they are decided first before building the
+  // windows so the windows can be properly cut around them.
+
+  // Get the pointers to the input columnar data ready
+  std::vector<int8_t const *> input_data;
```

Comment:
I recently discovered a more concise phrasing for this:
```c++
auto data_begin = thrust::make_transform_iterator(tbl.begin(),
                                                  [](auto const& c) { return c.template data<int8_t>(); });
std::vector<int8_t const&> input_data( data_begin, data_begin + tbl.num_columns() );
```
A transforming constructor thingie.

Reply:
Similarly, `input_nm`.

Reply:
I attempted this, but couldn't get the compiler happy with it. It seemed to want to call the copy constructor, which is fine for a pointer, but not so great for a `rmm::device_buffer`. I left the change commented out for now so I can investigate further and also in case you wanted to try it out. I'm going to push a new commit with that removed though so this PR can continue the process, so be sure to check out the original commit if this interests you.